apitools/base/py/credentials_lib.py

```diff
@@ -48,65 +48,57 @@ def SetCredentialsCacheFileLock(lock):
     cache_file_lock = lock
 
 
-# TODO(craigcitro): Expose the extra args here somewhere higher up,
-# possibly as flags in the generated CLI.
+# List of additional methods we use when attempting to construct
+# credentials. Users can register their own methods here, which we try
+# before the defaults.
+_CREDENTIALS_METHODS = []
+
+
+def RegisterCredentialsMethod(method, position=None):
+    """Register a new method for fetching credentials.
+
+    This new method should be a function with signature:
+      client_info, **kwds -> Credentials or None
+    This method can be used as a decorator, unless position needs to
+    be supplied.
+
+    Note that method must *always* accept arbitrary keyword arguments.
+
+    Args:
+      method: New credential-fetching method.
+      position: (default: None) Where in the list of methods to
```

Comment:
If the caller needs to control it to that extent, what do you think about returning a list of methods to the caller and requiring them to rebuild with their own order, possibly disincluding certain methods? Something along the lines of this, from the caller's perspective:

default_methods = credentials_lib.GetCredentialsMethods()
get_creds_methods = []
get_creds_methods.append(default_methods['GetServiceAccountCredentials'])
get_creds_methods.append(_UserSpecifiedCredentialsMethod)
for method_name, method in default_methods.iteritems():
  if (method_name != 'GetServiceAccountCredentials' and
      method_name != 'GetApplicationDefaultCredentials'):
    get_creds_methods.append(method)

credentials_lib.SetCredentialsMethods(get_creds_methods)

I don't think we can make it easy on the caller when apitools adds a new method, but at the least we could make the ordering explicit and centralized.